### PR TITLE
API 5.0 WP5 - Pinned messages

### PR DESCRIPTION
**Pull request by harshil21**
Along with the bot API update, I also added a `unpin()` shortcut to message.py and `unpin_all_chat_messages()` shortcut to chat.py.
Also if the tests I wrote aren't good enough, please feel free to edit them.

The mypy build is failing because of something webhook handler related (probably due to API 5.0) (edit: Nvm it only failed for me locally, it's not failing here)

### CODE REVIEW

File: telegram/bot.py [hunk 1]
```diff
@@ -3664,17 +3664,23 @@ def pin_chat_message(
 
     @log
     def unpin_chat_message(
-        self, chat_id: Union[str, int], timeout: float = None, api_kwargs: JSONDict = None
+        self,
+        chat_id: Union[str, int],
+        message_id: Union[str, int] = None,
```
Comment (Bibo-Joshi) [inline telegram/bot.py:3669]: please put the new parameter last, just in case someone uses positional args

File: telegram/bot.py [hunk 2]
```diff
@@ -3690,10 +3696,49 @@ def unpin_chat_message(
...
+            :obj:`bool`: On success, :obj:`True` is returned.
+
+        Raises:
+            :class:`telegram.TelegramError`
+
+        """
+
+        data: JSONDict = {'chat_id': chat_id}
+
+        result = self._post('unpinAllChatMessages', data, timeout=timeout, api_kwargs=api_kwargs)
```
Comment (Bibo-Joshi) [inline telegram/bot.py:3738]: you can just `return self._post(…)`

File: telegram/chat.py [hunk 3]
```diff
@@ -277,6 +277,19 @@ def set_administrator_custom_title(self, *args: Any, **kwargs: Any) -> bool:
         """
         return self.bot.set_chat_administrator_custom_title(self.id, *args, **kwargs)
 
+    def unpin_all_chat_messages(self, *args: Any, **kwargs: Any) -> bool:
```
Comment (Bibo-Joshi) [inline telegram/chat.py:280]: can you add this to `User`, too, as we now can pin in private chats?

File: tests/test_bot.py [hunk 4]
```diff
@@ -1263,15 +1263,22 @@ def test_set_chat_description(self, bot, channel_id):
...
-        message = bot.send_message(super_group_id, text="test_pin_message")
+        message1 = bot.send_message(super_group_id, text="test_pin_message_1")
+        message2 = bot.send_message(super_group_id, text="test_pin_message_2")
+        message3 = bot.send_message(super_group_id, text="test_pin_message_3")
+
         assert bot.pin_chat_message(
-            chat_id=super_group_id, message_id=message.message_id, disable_notification=True
+            chat_id=super_group_id, message_id=message1.message_id, disable_notification=True
         )
+        message2.pin()
```
Comment (Bibo-Joshi) [inline tests/test_bot.py:1273]: please use the plain bot methods here, testing the shortcuts should be done in the respective tests ;)

File: tests/test_chat.py [hunk 5]
```diff
@@ -183,6 +183,13 @@ def test(*args, **kwargs):
         monkeypatch.setattr('telegram.Bot.set_chat_administrator_custom_title', test)
         assert chat.set_administrator_custom_title(42, 'custom_title')
 
+    def test_unpin_all_chat_messages(self, monkeypatch, chat):
+        def test(*args, **kwargs):
```
Comment (Bibo-Joshi) [inline tests/test_chat.py:187]: I recently took the habit of giving these internal functions a more speaking name :D maybe use `assertion` or `make_assertion`

File: tests/test_message.py [hunk 6]
```diff
@@ -1052,6 +1052,15 @@ def test(*args, **kwargs):
         monkeypatch.setattr(message.bot, 'pin_chat_message', test)
         assert message.pin()
 
+    def test_unpin(self, monkeypatch, message):
+        def test(*args, **kwargs):
```
Comment (Bibo-Joshi) [inline tests/test_message.py:1056]: same

File: tests/test_bot.py [hunk 7]
```diff
@@ -1263,15 +1263,22 @@ def test_set_chat_description(self, bot, channel_id):
...
+        message3.pin()
```
Comment (Bibo-Joshi) [inline tests/test_bot.py:1274]: using `bot.pin_chat_message` without passing `message_id` should also cover the uncovered line marked by codecov

**Review by Bibo-Joshi — CHANGES_REQUESTED** [review]
Thanks for the PR! Left some notes :)

File: telegram/chat.py (hunk 3 above)
Comment (harshil21) [reply telegram/chat.py:280]: Oh right, will do

File: tests/test_bot.py (hunk 7 above)
Comment (harshil21) [reply tests/test_bot.py:1274]: uhm, how will the message be pinned without the `message_id`? Isn't it required? Did you mean `unpin_chat_message`?

File: telegram/bot.py (hunk 2 above)
Comment (harshil21) [reply telegram/bot.py:3738]: That `result = ...` style is being used everywhere else too. Change those occurrences too?

Comment (Bibo-Joshi) [reply telegram/bot.py:3738]: no need, it's not too important. just noticed here, so we can change it here :)

File: telegram/chat.py (hunk 3 above)
Comment (harshil21) [reply telegram/chat.py:280]: Should I also add `pin()`, `unpin()` and `unpin_all()` shortcuts in `CallbackQuery`?

Comment (Bibo-Joshi) [reply telegram/chat.py:280]: `pin` and `unpin` make sense, `unpin_all` not reall imo.

File: tests/test_bot.py (hunk 7 above)
Comment (Bibo-Joshi) [reply tests/test_bot.py:1274]: uh, yeah, mixed that up, sry. anyway, that if-clause branch in `unpit_chat_message` should be tested :D

File: telegram/chat.py (hunk 3 above)
Comment (Bibo-Joshi) [reply telegram/chat.py:280]: uh, please call them `(un)_pin_message`, as we're not pinning the callback query ;)

File: telegram/bot.py (hunk 2 above)
Comment (harshil21) [reply telegram/bot.py]: the `# type: ignore[return-value]` thing doesn't work after doing that (hence failing tests), am I doing it wrong or should I revert?

Comment (harshil21) [reply telegram/bot.py]: Nvm fixed it, that thing is weird

File: telegram/user.py [hunk 8]
```diff
@@ -189,6 +189,45 @@ def mention_html(self, name: str = None) -> str:
             return util_mention_html(self.id, name)
         return util_mention_html(self.id, self.full_name)
 
+    def pin_message(self, *args: Any, **kwargs: Any) -> bool:
```
Comment (Bibo-Joshi) [inline telegram/user.py:192]: why does `User` get `un/pin_message` shortcuts and `Chat` doesn't? :D Tbh I don't really see a big use case for `User/Chat.un/pin_message`, but having doesn't hurt. In any case, either both should have them or none of them :)

**Review by Bibo-Joshi — CHANGES_REQUESTED** [review]
Thanks for the updates. Left some more nitpicking. Don't mind the pytest failures for now, TG seems to need a break from us (hopefully that will get better once we use a selfhosted API instance 😎 ). Codecov is still failing, though:

image

The line marked in read is not hit by the tests

**Comment by harshil21** [comment]
> The line marked in read is not hit by the tests

Yea just realized that I was testing that as a positional argument 😳

File: telegram/user.py (hunk 8 above)
Comment (harshil21) [reply telegram/user.py:192]: Yea my bad, I was being lazy. Changed in the next commit

**Comment by harshil21** [comment]
Finally all tests pass. Had to redo some tests cause I don't understand when/why *args/**kwargs is empty in the monkeypatch function

File: telegram/callbackquery.py [hunk 9]
```diff
@@ -325,3 +325,29 @@ def delete_message(self, *args: Any, **kwargs: Any) -> Union[Message, bool]:
 
         """
         return self.message.delete(*args, **kwargs)
+
+    def pin_message(self, *args: Any, **kwargs: Any) -> bool:
+        """Shortcut for::
+
+             bot.pin_chat_message(chat_id=update.callback_query.message.chat_id,
+                                  *args,
```
[the remark this answers is not in the document]
Comment (Bibo-Joshi) [reply telegram/callbackquery.py:333]: Please also update the docstring (also for `unpin`) :)

**Review by Bibo-Joshi — CHANGES_REQUESTED** [review]
Just one minor nitpick left from my side. @Poolitzer will want to review as well, though.

File: telegram/bot.py [hunk 10]
```diff
@@ -3690,9 +3696,48 @@ def unpin_chat_message(
...
+        self,
+        chat_id: Union[str, int],
+        timeout: float = None,
+        api_kwargs: JSONDict = None,
+    ) -> bool:
+        """
+        Use this method to remove a message from the list of pinned messages in a chat. If the
+        chat is not a private chat, the bot must be an administrator in the chat for this
+        to work and must have the ``can_pin_messages`` admin right in a supergroup or
+        ``can_edit_messages`` admin right in a channel.
```
Comment (Poolitzer) [inline telegram/bot.py:3717]: This description is wrong

File: telegram/callbackquery.py [hunk 11]
```diff
@@ -325,3 +325,29 @@ def delete_message(self, *args: Any, **kwargs: Any) -> Union[Message, bool]:
...
+
+    def unpin_message(self, *args: Any, **kwargs: Any) -> bool:
+        """Shortcut for::
+
+             bot.unpin_chat_message(chat_id=message.chat_id,
+                                    *args,
+                                    **kwargs)
+
+        Returns:
+            :obj:`bool`: On success, :obj:`True` is returned.
```
Comment (Poolitzer) [inline telegram/callbackquery.py:350]: you also set the message id with the call you do later.

**Review by Poolitzer — CHANGES_REQUESTED** [review]
>  @Poolitzer will want to review as well, though.

Dam Dam Daaaam. He wants though yeah. And he did some nitpicking about docs. And talks about himself in the third perspective.

Weird guy.

**Review by Bibo-Joshi — CHANGES_REQUESTED** [review]
I just noticed that the docsring of `Message.pinned_message` changed as well, please update that :)

File: telegram/bot.py (hunk 10 above)
Comment (harshil21) [reply telegram/bot.py:3717]: :O can't believe I didn't catch that

**Comment by harshil21** [comment]
> I just noticed that the docsring of `Message.pinned_message` changed as well, please update that :)

It didn't?
image

File: telegram/bot.py (hunk 10 above)
Comment (Poolitzer) [reply telegram/bot.py:3717]: thats what I am here for no worries

**Comment by Poolitzer** [comment]
image
It did for me

**Comment by harshil21** [comment]
> image
> It did for me

That's in `Chat.pinned_message`, which I did update. The one in `Message.pinned_message` didn't change though

**Comment by Poolitzer** [comment]
oh thats right. Pretty sure that bibo mixed them up as well

**Comment by Bibo-Joshi** [comment]
ups, my bad 😬  sorry for the confusion

**Comment by Bibo-Joshi** [comment]
For some reason codecov marks docstring as uncovered, but afais all change requests were resolved. merging.
thanks for the contribution :)

**Pull request merged**